### ISSUE TITLE
s4 not working within testthat environment

### ISSUE THREAD

**Issue by mikejiang** [r-lib/testthat#811]
My `flowClust` package imports s4 generic `Map` from `BiocGenerics` and exports it as s4 method
https://github.com/RGLab/flowClust/blob/trunk/NAMESPACE#L39-L51
```r
exportMethods(Map)
importFrom(BiocGenerics,Map)
```
As shown below, S4 generic `Map` masks `base::Map` as expected after loading `flowClust` under the regular R environment
```r
> Map
function (f, ...) 
{
    f <- match.fun(f)
    mapply(FUN = f, ..., SIMPLIFY = FALSE)
}
<bytecode: 0x31ccd38>
<environment: namespace:base>
> library(flowClust)
Attaching package: ‘flowClust’
> Map
standardGeneric for "Map" defined from package "BiocGenerics"

function (f, ...) 
standardGeneric("Map")
<bytecode: 0x6090ba0>
<environment: 0x6089498>
Methods may be defined for arguments: ...
Use  showMethods("Map")  for currently available ones.
```

However,  it doesn't work in the test case of another package `CytoML`, which tries to call `flowClust::Map` S4 method while `BiocGenerics::Map` remains invisible thus fails to dispatch the S4 method.

```r
> devtools::test(filter = "flowJo")
Loading CytoML
Loading required package: testthat
Testing CytoML
✔ | OK F W S | Context
Called from: eval(code, test_env)
Browse[1]> Map
function (f, ...) 
{
    f <- match.fun(f)
    mapply(FUN = f, ..., SIMPLIFY = FALSE)
}
<bytecode: 0x31ccd38>
<environment: namespace:base>
```

Apparently, `flowClust` is preloaded in [`helper-` file](https://github.com/RGLab/CytoML/blob/test/tests/testthat/helper-lib.R). Btw, I haven't had this issue until recently.
Here is the [sessionInfo.txt](https://github.com/r-lib/testthat/files/2531162/sessionInfo.txt)
Here is the [test case](https://github.com/RGLab/CytoML/blob/test/tests/testthat/test-flowJo.R#L86-L93)

**Comment by maxheld83** [comment]
Is this perhaps related to another issue I just encountered/raised in [devtools](https://github.com/r-lib/devtools/issues/1917): It appeared to me that I couldn't *call* S3 methods in `tests/testthat/helper_*.R`-files, because those (starting with devtools >= 2.0.0) would bork up devtools::document().

**Comment by jacobpwagner** [comment]
Thanks @maxheld83! You appear to be right that it's related to https://github.com/r-lib/devtools/issues/1917. I used your fix from 4ca497496789965a969b36b98fa09cef256d1ec1 and it worked like a charm.

**Comment by hadley** [comment]
This is mostly likely to be something you need to tweak on your end. Unfortunately the interaction between S4 and environments is complicated, and we're not S4 experts.
